lib/composite_primary_keys/composite_predicates.rb

```diff
@@ -52,40 +52,55 @@ def cpk_join_predicate(table1, key1, table2, key2)
 
     def cpk_in_predicate(table, primary_keys, ids)
       if primary_keys.length == 2
-        low_cardinality_key_part, high_cardinality_key_part = if ids.map(&:second).uniq.size > ids.map(&:first).uniq.size
-                                                                %i[first second]
-                                                              else
-                                                                %i[second first]
-                                                              end
-
-        groups = ids.group_by(&low_cardinality_key_part).transform_values do |values|
-          values.map(&high_cardinality_key_part)
+        cpk_in_predicate_with_grouped_keys(table, primary_keys, ids)
+      else
+        cpk_in_predicate_with_non_grouped_keys(table, primary_keys, ids)
+      end
+    end
+
+    def cpk_in_predicate_with_non_grouped_keys(table, primary_keys, ids)
+      and_predicates = ids.map do |id|
+        cpk_id_predicate(table, primary_keys, id)
+      end
+
+      cpk_or_predicate(and_predicates)
+    end
+
+    def cpk_in_predicate_with_grouped_keys(table, primary_keys, ids)
+      keys_by_first_key_part = Hash.new { |h, k| h[k] = [] }
+      keys_by_second_key_part = Hash.new { |h, k| h[k] = [] }
+
+      ids.map.each do |first_key_part, second_key_part|
+        keys_by_first_key_part[first_key_part] << second_key_part
+        keys_by_second_key_part[second_key_part] << first_key_part
+      end
+
+      low_cardinality_key_part, high_cardinality_key_part, groups = \
+        if keys_by_first_key_part.keys.size <= keys_by_second_key_part.keys.size
+          primary_keys + [keys_by_first_key_part]
```

Comment:
I'm confused by this. Why is primary keys being added to the hash table?

Reply:
This code is making an array where the elements are `[primary_keys.first, primary_keys.second, groups]` (not modifying the hash). When keys_by_second_column_name has fewer keys the returned array is `[primary_keys.second, primary_keys.first, groups]`. This is important to make sure that `low_cardinality_column_name` and `high_cardinality_column_name` are bound to the right column_names.

I'm thinking that maybe explicitly defining the arrays rather than calculating them might be more clear. I'll push that up and you can see what you think.